source/common/config/version_converter.cc

```diff
@@ -40,5 +94,59 @@ DowngradedMessagePtr VersionConverter::downgrade(const Protobuf::Message& messag
   return downgraded_message;
 }
 
+DynamicMessagePtr
+VersionConverter::reinterpret(const Protobuf::Message& message,
+                              envoy::config::core::v3alpha::ApiVersion api_version) {
+  switch (api_version) {
+  case envoy::config::core::v3alpha::ApiVersion::AUTO:
+  case envoy::config::core::v3alpha::ApiVersion::V2:
+    // TODO(htuch): this works as long as there are no new fields in the v3+
+    // DiscoveryRequest. When they are added, we need to do a full v2 conversion
+    // and also discard unknown fields. Tracked at
+    // https://github.com/envoyproxy/envoy/issues/9619.
+    return downgrade(message);
```

Comment:
Sorry for being dense but I'm still a little confused around all of this. Is it possible to add some more comments throughout these functions with possibly some simple examples? If I'm downgrading from v3 -> v2, but I have a v3 message that has a smuggled deprecated field, how does that work with the serialize as string conversion? The old message doesn't know about the deprecated rename, so wouldn't we need code to actually map it back?

Reply:
I can add some, but the key intuition is that v2/v3 actually have exactly the same fields from a proto wire perspective if we are using the v3 shadow protos. They are just different names, which proto doesn't care about on the wire.

Reply:
Ah OK sorry, I saw `SerializeAsString` and I was thinking about a debug string which would have field names. We are just going back and forth with bytes. OK got it. I think some more comments would be good but feel free to do as a follow up.